RecoLocalTracker/SiStripZeroSuppression/plugins/SiStripZeroSuppression.cc

```diff
@@ -164,7 +164,7 @@ inline edm::DetSet<SiStripRawDigi> SiStripZeroSuppression::formatRawDigis(const
   outRawDigis.reserve(rawDigis.size());
   const std::vector<bool>& apvf = algorithms->getAPVFlags();
   uint32_t strip = 0;
-  for (const auto rawDigi : rawDigis) {
+  for (const auto& rawDigi : rawDigis) {
```

Comment:
I am not fully up to date with the latest recommendation, but why is this a warning? SiStripRawDigi (and SiStripDigi, in the two changes below) are small (one or two `uint16_t`, respectively), so the code was intentionally written like this. Or will the compiler pick the most efficient option in any case when using `const auto&`?

Reply:
I think regardless of the size of object, LLVM 10 complains that you are creating a copy of object and suggests to use a `&`

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc820/CMSSW_11_1_CLANG_X_2020-04-27-2300/RecoLocalTracker/SiStripZeroSuppression
```
  RecoLocalTracker/SiStripZeroSuppression/plugins/SiStripZeroSuppression.cc:167:19: warning: loop variable 'rawDigi' of type 'const SiStripRawDigi' creates a copy from type 'const SiStripRawDigi' [-Wrange-loop-construct]
   for (const auto rawDigi : rawDigis) {
                  ^
RecoLocalTracker/SiStripZeroSuppression/plugins/SiStripZeroSuppression.cc:167:8: note: use reference type 'const SiStripRawDigi &' to prevent copying
  for (const auto rawDigi : rawDigis) {
       ^~~~~~~~~~~~~~~~~~~~
                  &
1 warning generated.
```

Reply:
hope the compiler will optimize either way

Reply:
IIRC, we have quite a few cases where a copy is OK for small types.
It would be good to understand the underlying logic why it's triggered here.

Reply:
Playing around with compiler explorer revealed that it is the non-default destructor of `SiStripRawDigi`
https://github.com/cms-sw/cmssw/blob/b3d42e1d090033076aec4580d2395837867124e6/DataFormats/SiStripDigi/interface/SiStripRawDigi.h#L20
makes clang to issue the warning. Using the default destructor either by
```cpp
  ~SiStripRawDigi() = default;
```
or removing the explicit definition removes the warning (while keeping the loop variable taken by value).

Reply:
According to Clang 10.0 release note [a], looks like clang is not treating simple class like SiStripRawDigi as small/trivially copyable types

[a]
https://releases.llvm.org/10.0.0/tools/clang/docs/ReleaseNotes.html#id3 ). So looks like clang 10.0 is not treating 
```
-Wrange-loop-analysis got several improvements. It no longer warns about a copy being
made when the result is bound to an rvalue reference. It no longer warns when an object
of a small, trivially copyable type is copied. The warning now offers fix-its. Excluding 
-Wrange loop-bind-reference it is now part of -Wall. To reduce the number of false
positives the diagnostic is disabled in macros and template instantiations.
```

Reply:
So in principle going for default destructors where possible would be a slightly better solution.

Reply:
thanks @makortel . 
@pieterdavid @slava77 ,  let me know if you would like me to revert this change and go for default destructors  for SiStripRawDigi class.

Reply:
it sounds like the default constructor is a better solution.
If the compiler were to take this loop as is, the reference solution is supposedly more costly on this type with uint16_t memory footprint. I'm not sure how effective is the compiler optimization to get rid of the reference completely.